src/engine-actions/get-payment-preimage.js

```diff
@@ -27,62 +20,82 @@ const POLLING_INTERVAL = 10000
 
-/**
- * Retrieve the preimage for a completed payment specified by the payment
- * hash.
- * @todo LND has a known issue preventing payments that are settled after a restart
- *       showing up in this list, which will mistakenly throw an error.
- *       {@link https://github.com/lightningnetwork/lnd/issues/2032}
- * @param  {string} paymentHash - Base64 encoded payment hash
- * @param  {Object} options
- * @param  {LndClient} options.client
- * @returns {string} Base64 encoded payment preimage associated with the hash
- * @throws {Error} If payment is not completed
- */
-async function getCompletedPreimage (paymentHash, { logger, client }) {
-  const { payments = [] } = await listPayments({ client })
+      const endListener = () => {
+        this.logger.error('LND closed trackPayment stream before returning our value')
+        return finish(new Error('LND closed stream to track payment'))
+      }
 
-  logger.debug(`Found ${payments.length} completed payments`)
+      const dataListener = (response) => {
+        const { state, preimage } = response
 
-  const paymentHashBuf = Buffer.from(paymentHash, 'base64')
-  const payment = payments.find(({ paymentHash }) => {
-    // `paymentHash` returned from ListPayments {@link https://api.lightning.community/#payment}
-    // is a hex string. We need to compare to our payment hash which is a base64 string,
-    // so we convert both to buffers to do the comparison.
-    return paymentHashBuf.equals(Buffer.from(paymentHash, 'hex'))
-  })
+        if (state === PAYMENT_STATUSES.IN_FLIGHT) {
+          // trackPayment will continue to stream updates of the payment state, so
+          // no action is required on our end
+          this.logger.debug('Payment is In-Flight for paymentHash:', { paymentHash })
+        }
+
+        if (state === PAYMENT_STATUSES.SUCCEEDED) {
+          if (!preimage) {
+            return finish(new Error('No preimage associated with successful payment'))
+          }
 
-  if (!payment) {
-    throw new Error(`No completed payment with hash '${paymentHash}' exists.`)
-  }
+          this.logger.debug('Payment has Succeeded for paymentHash', { paymentHash })
+          return finish(null, { paymentPreimage: preimage })
+        }
 
-  // like the `paymentHash`, the `paymentPreimage` is returned as a hex string from LND,
-  // so we convert to Base64 before returning
-  return Buffer.from(payment.paymentPreimage, 'hex').toString('base64')
+        if (state === PAYMENT_STATUSES.FAILED_TIMEOUT || state === PAYMENT_STATUSES.FAILED_NO_ROUTE) {
+          this.logger.error('Payment has failed', { paymentHash })
+          const permanentError = new Error(`Payment failed for paymentHash: ${paymentHash}`)
+          return finish(null, { permanentError })
+        }
+
+        throw new Error(`Unknown payment status ${state} for payment with hash '${paymentHash}'`)
+      }
+
+      const finish = (err, response) => {
+        stream.removeListener('error', errorListener)
```

Comment:
`stream.removeAllListeners()` would work here instead of going one by one. It might also let you `reject` or `resolve` in the specific handlers instead of defining a separate `finish` function.

Reply:
Let me know what you think of the style with `finish` removed. Redundant calls to `stream.removeAllListeners()` isn't my favorite, but no biggie either way.